File: Cplusplus/0054_spiral_matrix/0054_spiral_matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> result;
        if (matrix.size() == 0)
            return result;
        int n = matrix.size();
        int m = matrix[0].size();
        int count = 1;
        int begin = 0;
        while (count <= m * n) {
            for (int i = begin; i < m - begin; i++) {
                result.push_back(matrix[begin][i]);
                count++;
                if(count > m * n)
                  return result;
            }

            for (int i = begin + 1; i < n - begin; i++) {
                result.push_back(matrix[i][m - 1 - begin]);
                count++;
                if(count > m * n)
                  return result;
            }

            for (int i = m - 2 - begin; i >= begin; i--) {
                result.push_back(matrix[n - 1 - begin][i]);
                count++;
                if(count > m * n)
                  return result;
            }

            for (int i = n - 2 - begin; i > begin; i--) {
                result.push_back(matrix[i][begin]);
                count++;
                if (count > m * n)
                  return result;
            }
            begin++;
        }
        return result;
    }
};

File: Python/0054_spiral_matrix/0054_spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        result = []
        if len(matrix) == 0:
            return result
        n = len(matrix)
        m = len(matrix[0])
        count = 1
        begin = 0
        while count <= m * n:
            for i in range(begin, m - begin):
                result.append(matrix[begin][i])
                count += 1
                if count > m * n:
                    return result
            for i in range(begin + 1, n - begin):
                result.append(matrix[i][m - 1 - begin])
                count += 1
                if count > m * n:
                    return result
            i = m - 2 - begin
            while i >= begin:
                result.append(matrix[n - 1 - begin][i])
                count += 1
                if count > m * n:
                    return result
                i -= 1
            i = n - 2 - begin
            while i > begin:
                result.append(matrix[i][begin]);
                count += 1
                if count > m * n:
                    return result
                i -= 1
            begin += 1
        return result

